### PR TITLE
feat: add opt out analytics

### PR DESCRIPTION
**Pull request by JackShort**
* Adds toggle to opt out of analytics
* adds wrappers around analytics functions to check atom value if analytics should be sent
Screen Shot 2023-07-19 at 4 47 00 PM

_I am sorry for the number of files this touches - but it if makes you feel better it sucked for me too_

### CODE REVIEW

File: src/components/AccountDrawer/SmallBalanceToggle.tsx [hunk 1]
```diff
@@ -1,30 +1,21 @@
-import { Trans } from '@lingui/macro'
-import Row from 'components/Row'
-import Toggle from 'components/Toggle'
+import { t } from '@lingui/macro'
```
Comment (zzmp) [inline src/components/AccountDrawer/SmallBalanceToggle.tsx:1]: nit: If you're trying to get it in faster, this refactor could be a separate PR.

File: src/components/analytics/TraceAnalytics.tsx [hunk 2]
```diff
@@ -0,0 +1,35 @@
+import { sendAnalyticsEvent, Trace, TraceEvent } from '@uniswap/analytics'
+import { atomWithStorage, useAtomValue } from 'jotai/utils'
+import { memo } from 'react'
+
+const analyticsOptOutAtomName = 'optOutOfAnalytics'
+export const analyticsOptOutToggle = atomWithStorage<boolean>(analyticsOptOutAtomName, true)
+
+export const TraceAnalytics = memo((props: React.ComponentProps<typeof Trace>) => {
```
Comment (zzmp) [inline src/components/analytics/TraceAnalytics.tsx:8]: This should be called `AnalyticsTrace`, not `TraceAnalytics`, following the filepath.

File: src/components/analytics/TraceAnalytics.tsx [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  const allowAnalytics = useAtomValue(analyticsOptOutToggle)
+
+  return <Trace {...props} shouldLogImpression={allowAnalytics} />
+})
+
+TraceAnalytics.displayName = 'TraceAnalytics'
+
+export const TraceAnalyticsEvent = memo((props: React.ComponentProps<typeof TraceEvent>) => {
```
Comment (zzmp) [inline src/components/analytics/TraceAnalytics.tsx:16]: This should be called `AnalyticsTraceEvent`, not `TraceAnalyticsEvent`, following the filepath. Alternatively, it could be called `AnalyticsEvent`.

File: src/components/analytics/TraceAnalytics.tsx [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  const allowAnalytics = useAtomValue(analyticsOptOutToggle)
+
+  return <TraceEvent {...props} shouldLogImpression={allowAnalytics} />
+})
+
+TraceAnalyticsEvent.displayName = 'TraceAnalyticsEvent'
+
+export const sendOptOutAnalyticsEvent: typeof sendAnalyticsEvent = (event, properties) => {
```
Comment (zzmp) [inline src/components/analytics/TraceAnalytics.tsx:24]: the "OptOut" is an implementation detail to the consumers of this file, so it should be omitted. Export this as `sendAnalyticsEvent`.

File: src/components/analytics/TraceAnalytics.tsx [hunk 5]
```diff
@@ -0,0 +1,35 @@
+import { sendAnalyticsEvent, Trace, TraceEvent } from '@uniswap/analytics'
+import { atomWithStorage, useAtomValue } from 'jotai/utils'
+import { memo } from 'react'
+
+const analyticsOptOutAtomName = 'optOutOfAnalytics'
+export const analyticsOptOutToggle = atomWithStorage<boolean>(analyticsOptOutAtomName, true)
```
Comment (zzmp) [inline src/components/analytics/TraceAnalytics.tsx:6]: Have we used other atoms with storage, or is this a new pattern? If we're using localStorage, should this potentially go through redux instead, to avoid writing to localStorage from multiple libraries?

File: src/components/analytics/TraceAnalytics.tsx [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+export const TraceAnalyticsEvent = memo((props: React.ComponentProps<typeof TraceEvent>) => {
+  const allowAnalytics = useAtomValue(analyticsOptOutToggle)
+
+  return <TraceEvent {...props} shouldLogImpression={allowAnalytics} />
+})
+
+TraceAnalyticsEvent.displayName = 'TraceAnalyticsEvent'
+
+export const sendOptOutAnalyticsEvent: typeof sendAnalyticsEvent = (event, properties) => {
+  const value = localStorage.getItem(analyticsOptOutAtomName)
```
Comment (zzmp) [inline src/components/analytics/TraceAnalytics.tsx:25]: Instead of referencing localStorage every time, could this be stored in-memory? Does it matter - ie, is accessing localStorage slow?

**Review by zzmp — CHANGES_REQUESTED** [review]
In order to prevent a regression, can you re-export `useTrace` from `components/analytics/TraceAnalytics`, and make a lint rule forbidding import of `@uniswap/analytics` except in `components/analytics/TraceAnalytics`?

I would also move the file from `components/analytics/TraceAnalytics` to `analytics`, as it is not just components, and it is very tightly scoped.

File: src/components/AccountDrawer/SmallBalanceToggle.tsx [hunk 7]
```diff
@@ -1,30 +1,21 @@
...
-            updateHideSmallBalances(!hideSmallBalances)
-          }}
-        />
-      </Row>
-    </Row>
+    <SettingsToggle
+      title={t`Hide small balances`}
+      isActive={hideSmallBalances}
+      toggle={() => {
+        updateHideSmallBalances(!hideSmallBalances)
```
Comment (zzmp) [inline src/components/AccountDrawer/SmallBalanceToggle.tsx:17]:
```suggestion
        updateHideSmallBalances((value) => !value)
```

File: src/components/AccountDrawer/SmallBalanceToggle.tsx [hunk 8]
```diff
@@ -1,30 +1,21 @@
...
+      }}
```
Comment (zzmp) [inline src/components/AccountDrawer/SmallBalanceToggle.tsx:18]:
```suggestion
      toggle={() => void updateHideSmallBalances((value) => !value)}
```

File: src/components/AccountDrawer/TestnetsToggle.tsx [hunk 9]
```diff
@@ -1,31 +1,22 @@
...
-        />
-      </Row>
-    </Row>
+    <SettingsToggle
+      title={t`Show testnets`}
+      dataid="testnets-toggle"
+      isActive={showTestnets}
+      toggle={() => {
+        updateShowTestnets(!showTestnets)
+      }}
```
Comment (zzmp) [inline src/components/AccountDrawer/TestnetsToggle.tsx:19]:
```suggestion
      toggle={() => void updateShowTestnets((value) => !value)}
```

File: src/components/AccountDrawer/AnalyticsToggle.tsx [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+  const [allowAnalytics, updateAllowAnalytics] = useAtom(analyticsOptOutToggle)
+
+  return (
+    <SettingsToggle
+      title={t`Allow analytics`}
+      description={t`We use anonymized data to enhance your experience with Uniswap Labs products.`}
+      isActive={allowAnalytics}
+      toggle={() => {
+        updateAllowAnalytics(!allowAnalytics)
+      }}
```
Comment (zzmp) [inline src/components/AccountDrawer/AnalyticsToggle.tsx:17]: This may cause race conditions if you double-click it. Instead, negate the current value - as opposed to the value in closure:

```suggestion
      toggle={() => voidupdateAllowAnalytics((value) => !value)}
```

File: src/components/analytics/TraceAnalytics.tsx (hunk 6 above)
Comment (JackShort) [reply src/components/analytics/TraceAnalytics.tsx:25]: its like a 10 thousandth of a millisecond to access local storage, so def not slow. The way to access this in-memory would be to use the useAtomValue hook in the component that calls this function then passing the value down, but that kinda defeats the purpose of this function

File: src/analytics/index.tsx [hunk 11]
```diff
@@ -0,0 +1,41 @@
+import {
+  sendAnalyticsEvent as sendAnalyticsTraceEvent,
+  Trace as AnalyticsTrace,
+  TraceEvent as AnalyticsEvent,
+} from '@uniswap/analytics'
+import { atomWithStorage, useAtomValue } from 'jotai/utils'
+import { memo } from 'react'
+
+const analyticsOptOutAtomName = 'optOutOfAnalytics'
+export const analyticsOptOutToggle = atomWithStorage<boolean>(analyticsOptOutAtomName, true)
```
Comment (zzmp) [inline src/analytics/index.tsx:10]: This reads like its negation. If you keep this name, you should invert the default value and all usages. Otherwise, something like 'allow_analytics'.

```suggestion
const analyticsOptOutKey = 'analytics_opt_out'
export const analyticsOptOutAtom = atomWithStorage<boolean>(analyticsOptOutKey, false)
```

File: src/analytics/index.tsx [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+
+export const Trace = memo((props: React.ComponentProps<typeof AnalyticsTrace>) => {
+  const allowAnalytics = useAtomValue(analyticsOptOutToggle)
+
+  return <AnalyticsTrace {...props} shouldLogImpression={allowAnalytics} />
```
Comment (zzmp) [inline src/analytics/index.tsx:15]:
```suggestion
  const analyticsOptOut = useAtomValue(analyticsOptOutAtom)
  const shouldLogImpression = analyticsOptOut ? false : props.shouldLogImpression
  return <AnalyticsTrace {...props} shouldLogImpression={shouldLogImpression} />
```

File: src/analytics/index.tsx [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+})
+
+Trace.displayName = 'Trace'
+
+export const TraceEvent = memo((props: React.ComponentProps<typeof AnalyticsEvent>) => {
+  const allowAnalytics = useAtomValue(analyticsOptOutToggle)
```
Comment (zzmp) [inline src/analytics/index.tsx:21]: See above.

File: src/analytics/index.tsx [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+
+export const sendAnalyticsEvent: typeof sendAnalyticsTraceEvent = (event, properties) => {
+  const value = localStorage.getItem(analyticsOptOutAtomName)
+  let canSendAnalytics = true
+
+  if (typeof value === 'string' && JSON.parse(value) === false) {
+    canSendAnalytics = false
+  }
+
+  if (canSendAnalytics) {
```
Comment (zzmp) [inline src/analytics/index.tsx:36]: This can still fail. Could you use a try/catch?

```suggestion
  let analyticsOptOut = false
  try {
    analyticsOptOut = JSON.parse(localStorage.getItem(analyticsOptOutKey)
  } catch { }

  if (!analyticsOptOut) {
```

File: src/analytics/index.tsx [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+    sendAnalyticsTraceEvent(event, properties)
+  }
+}
+
+export { getDeviceId, initializeAnalytics, OriginApplication, user, useTrace } from '@uniswap/analytics'
```
Comment (zzmp) [inline src/analytics/index.tsx:41]: nit: I prefer re-exports directly under imports.

File: src/components/Web3Provider/index.test.tsx [hunk 16]
```diff
@@ -15,7 +15,10 @@ import { mocked } from 'test-utils/mocked'
 
 import Web3Provider from '.'
 
-jest.mock('@uniswap/analytics', () => ({
+jest.mock('analytics', () => ({
+  user: { set: jest.fn(), postInsert: jest.fn() },
+}))
```
Comment (zzmp) [inline src/components/Web3Provider/index.test.tsx:20]:
```suggestion
```

**Review by zzmp — CHANGES_REQUESTED** [review]

File: src/analytics/index.tsx [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+
+TraceEvent.displayName = 'TraceEvent'
+
+export const sendAnalyticsEvent: typeof sendAnalyticsTraceEvent = (event, properties) => {
+  let allowAnalytics = true
+
+  try {
+    const value = localStorage.getItem(allowAnalyticsAtomKey)
+
+    if (typeof value === 'string') {
```
Comment (zzmp) [inline src/analytics/index.tsx:38]: You no longer need this check because you are in a try catch block.

**Review by zzmp — APPROVED** [review]
One last nit around checking typeof but lgtm

File: src/analytics/index.tsx (hunk 17 above)
Comment (JackShort) [reply src/analytics/index.tsx:38]: it actually won't let me call `JSON.parse` without checking because it could be `null`

**Review by cbachmeier — APPROVED** [review]
LGTM. Is there a way we can verify that analytic events are still being sent?

File: src/analytics/index.tsx [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+      allowAnalytics = JSON.parse(value)
+    }
+    // eslint-disable-next-line no-empty
+  } catch {}
+
```
Comment (cbachmeier) [inline src/analytics/index.tsx:43]: You could just console.error that value was null to remove this eslint ignore

Comment (JackShort) [reply src/analytics/index.tsx:43]: The problem is that will log it into sentry and idk if we want this to be logged to sentry because it's not an actual error

**Comment by JackShort** [comment]
> LGTM. Is there a way we can verify that analytic events are still being sent?

@cbachmeier actually one of the service worker tests checks if analytics is logged on first load
